sync/syncer.go

```diff
@@ -198,89 +216,191 @@ func (s *Syncer) getLayerFromNeighbors(currenSyncLayer types.LayerID) (*types.La
 	return types.NewExistingLayer(types.LayerID(currenSyncLayer), blocksArr), nil
 }
 
-func (s *Syncer) fetchFullBlocks(blockIds []types.BlockID) ([]*types.Block, error) {
-	output := s.fetchWithFactory(BlockReqFactory(blockIds), s.Concurrency)
+func (s *Syncer) FetchFullBlocks(blockIds []types.BlockID) ([]*types.Block, error) {
+	output := s.fetchWithFactory(NewBlockWorker(s, s.Concurrency, BlockReqFactory(), blockSliceToChan(blockIds)))
 	blocksArr := make([]*types.Block, 0, len(blockIds))
 	for out := range output {
 		block := out.(*types.Block)
-
-		eligible, err := s.BlockEligible(&block.BlockHeader)
-		if err != nil {
-			s.Warning(fmt.Sprintf("failed checking eligiblety %v", block.ID()), err)
-			continue
-		}
-		if !eligible {
-			s.Warning(fmt.Sprintf("block %v not eligible", block.ID()), err)
-			continue
-		}
-
-		if err := s.syncMissingContent(block); err != nil {
-			s.Error(fmt.Sprintf("failed derefrencing block data %v", block.ID()), err)
+		if err := s.SyncAndValidate(block); err != nil {
+			s.Error(fmt.Sprintf("failed derefrencing block data %v %v", block.ID(), err))
 			continue
 		}
 
 		s.Info("added block to layer %v", block.Layer())
 		blocksArr = append(blocksArr, block)
-
 	}
 
 	return blocksArr, nil
 }
 
-func (s *Syncer) syncMissingContent(blk *types.Block) error {
+func (s *Syncer) SyncAndValidate(blk *types.Block) error {
+	s.Info("sync and validate block %v ", blk.ID())
+
+	//check if known
+	if _, err := s.GetBlock(blk.Id); err == nil {
+		s.Info("we already know this block %v ", blk.ID())
+		return nil
+	}
+
+	//check block signature and is identity active
+	if eligable, err := s.IsIdentityActive(blk.MinerID.Key, blk.Layer()); err != nil || !eligable {
+		return errors.New(fmt.Sprintf("block %v identity activation check failed ", blk.ID()))
+	}
+
+	//associated atx data availability
+	if blk.ATXID != *types.EmptyAtxId {
+		if _, err := s.syncAtxs([]types.AtxId{blk.ATXID}); err != nil {
```

Comment:
You should validate that the ATX is for the current epoch

Reply:
IsIdentityActive() already checks, I hope, that the associated ATX (exists) and valid. That is, if the associated ATX is not valid, we won't get here.

We should make sure that the associated ATX belongs to this miner (generally, we should make sure that the associated ATX is the previous epoch).

In fact, we don't need the IsIdentityActive() function:
1. we should fetch the AssATX, and check that its miner's (Ed) id equals to the extracted pk.
2. we should check that the AssATX belongs to the previous epoch.
(I've opened a bug on IsIdentityActive() )

Reply:
`IsIdentityActive` only receives the block's layer and miner's ID, so it can't validate the associated ATX. Hence we need to validate that the associated ATX is - a. known to us, b. for the epoch in which the block was produced (not necessarily received) and c. that it belongs to the miner.

We shouldn't check only the associated ATX since we also expect that the ATX was reported in a past block (what you're suggesting is to treat all ATX's as free-standing)

Reply:
cool ill change it

Reply:
@barakshani we said we wont change IsIdentityActive  as part of this pr

Reply:
lets clarify exactly what needs to happen as part of this pr so we don't get no surprises later on 

@barakshani  "we should fetch the AssATX" like we said we are not fetching just checking local database 

regarding the bug you talked about dose it affect this pr ?